<a href="https://colab.research.google.com/github/Jeelpatel192/Python_Programming_Lab_Manual/blob/main/Python_Lab.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Program : 1

**Write a program to demonstrate basic data types and purposefully raise indentation error along with its correction.**

In [ ]:
a = 10
print(type(a))

<class 'int'>


In [ ]:
b = 2.13
print(type(b))

<class 'float'>


In [ ]:
c = 5 + 6j
print(type(c))

<class 'complex'>


In [ ]:
d = ("Hello")
print(type(d))

<class 'str'>


In [ ]:
e = [10,"Patel"]
print(type(e))

<class 'list'>


In [ ]:
f = {"Hello","World"}
print(type(f))

<class 'set'>


In [ ]:
g = dict(Name = "Jazz",Surname = "Patel")
print(type(g))

<class 'dict'>


In [ ]:
i = True
print(type(i))

<class 'bool'>


In [ ]:
j = False
print(type(j))

<class 'bool'>


In [ ]:
v = tuple(("Hiii"))
print(type(v))

<class 'tuple'>


In [ ]:
k = {'bye',12}
print(type(k))

<class 'set'>


# Program : 2

**Write a program to compute distance between two points taking input from the user (Pythagorean Theorem).**

In [ ]:
x1 = int(input("Enter x1 : "))
y1 = int(input("Enter y1 : "))
x2 = int(input("Enter x2 : "))
y2 = int(input("Enter y2: "))

result = ((((x2 - x1)**2) + ((y2 - y1)**2))**0.5)
#result = math.sqrt((x1 - x2)**2 + (y1 - y2)**2)

print("distance between",(x1,y1),"and",(x2,y2),"is : ",result)

Enter x1 : 5
Enter y1 : 6
Enter x2 : 2
Enter y2: 3
distance between (5, 6) and (2, 3) is :  4.242640687119285


# Program : 3

**Write a program
(a) For checking whether the given number is even or not
(b) To check whether given number is Armstrong or not.**

**(a) For checking whether the given number is even or not**

In [ ]:
num = int(input("Enter any numbers to test whether it is odd or even : "))

if(num % 2) == 0:
  print("The number is even")

else:
  print("The number is odd")

Enter any numbers to test whether it is odd or even : 5
The number is odd


**(b) To check whether given number is Armstrong or not.**

In [ ]:
n = int(input("Enter a number : "))

s = 0
t = n
while t > 0:
  digit = t % 10
  s += digit ** 3
  t //= 10

if n == s:
  print(n,"is an Armstrong number")

else:
  print(n,"is not an Armstrong number")


Enter a number : 123
123 is not an Armstrong number


# Program : 4

**Write a program using a while loop that asks the user for a number and prints a countdown from that number to zero.**

In [ ]:
a = int(input("Enter the Number : "))
i = a

while i >= 0:
  print(i)
  i = i - 1

Enter the Number : 5
5
4
3
2
1
0


# Program : 5

**Write a program to count the number of characters in the string and store them in a dictionary data structure.**

In [ ]:
a = str(input("Enter a string : "))
b = { }
for i in a:
  b[i] = a.count(i)
print(b)

Enter a string : Hello
{'H': 1, 'e': 1, 'l': 2, 'o': 1}


# Program : 6

**Write a program to demonstrate list and tuple in python. Write a program using a for loop that loops over a sequence.**

In [ ]:
list = ["Path","Python"]
print(list)

tuple = ("cotlin","java")
print(tuple)

for i in list:
  print(i)
for i in tuple:
  print(i)

['Path', 'Python']
('cotlin', 'java')
Path
Python
cotlin
java


# Program : 7

**Write a python program that should perform the following four tasks: (1) After getting a word (of input) from the user, your program should use a while (or for) loop to print outreach of the letters of the word. just remember that strings in python start with element 0!  (2) Your program should then use another loop to print out each of the letters of the (same) word in reverse order.	(3) Make a new variable that is the original word in reverse and print that variable. you can do this easily in python with a “string slice”. (4) Ask the user for a letter to count. Use another loop to count how many times that letter appears in the original word. print out this count.**

**1. After getting a word (of input) from the user, your program should use a while (or for) loop to print out each of the letters of the word. Just remember that strings in Python start with element 0!**

In [ ]:
str = input ("Enter the Word :- ")
if str [ : 2] == "0!":
    for i in range (len(str)) :
        print (str[i])
else :
    print ("Please enter again ")

Enter the Word :- String
Please enter again 


**2. Your program should then use another loop to print out each of the letters of the (same) word in reverse order!**

In [ ]:
str = input ("Enter the Word :- ")
for i in range (-1, -len(str) - 1, -1) :
    print (str[i])

Enter the Word :- Good
d
o
o
G


**3. Make a new variable that is the original word in reverse and print that variable. You can do this very easily in Python with a "string slice".**

In [ ]:
str = input ("Enter the Word :- ")
newstr = str[-1 : : -1]
print (newstr)

Enter the Word :- Karnavati
itavanraK


**4. Ask the user for a letter to count. Use another loop to count how many times that letter appears in the original word. Print out this count.**

In [ ]:
str = input ("Enter the Word :- ")
letter = input ("Enter a letter to count :- ") 
count = 0
for i in str :
    if i == letter :
        count += 1
print ("Number of letter :- ", count )

Enter the Word :- Hello
Enter a letter to count :- l
Number of letter :-  2


# Program : 8

**Write a program to print the mean, median, mode and range of the list.**

**Mean**

In [ ]:
num = [1, 2, 3, 4, 5]
n = len(num)
  
get_sum = sum(num)
mean = get_sum / n
  
print("Mean / Average is: " + str(mean))

Mean / Average is: 3.0


**Median**

In [ ]:
num = [1, 2, 3, 4, 5]
n = len(num)
num.sort()
  
if n % 2 == 0:
    median1 = num[n//2]
    median2 = num[n//2 - 1]
    median = (median1 + median2)/2
else:
    median = num[n//2]
print("Median is: " + str(median))

Median is: 3


**Mode**

In [ ]:
y= [11, 8, 8, 3, 4, 4, 5, 6, 6, 6, 7, 8]
y.sort()
L1=[]
i = 0
while i < len(y) :
    L1.append(y.count(y[i]))
    i += 1
d1 = dict(zip(y, L1))
d2={k for (k,v) in d1.items() if v == max(L1) }
print("mode is ",(d2))

mode is  {8, 6}
mode is  {8, 6}


**Range**

In [ ]:
nums = [9, 4, 6, 6, 5, 2, 10, 12, 1, 4, 4, 6]

minimum = min(nums)
maximum = max(nums)

print(f'The range is {minimum}, {maximum}')

The range is 1, 12


#Program : 9

**Write a program that uses a dictionary that contains ten usernames and passwords. the program should ask the user to enter their usernames and passwords. if the username is not in the dictionary, the program should indicate that the person is not a valid user of the system. if the username is in the dictionary, but the user does not enter the right password, the program should say that the password is invalid. if the password is correct, then the program should tell the user that they are now logged into the system.**

In [ ]:
dic = {'Path' : "2003",
       'walla' : "1809",
       'Portal' : "1912",
       'Express' : "2003",
       'Computer' : "9834",
       'Python' : "1990",
       'Mysql' : "2001",
       'c++' : "1996",
       'java' : "1998",
       'html' : "1999" }

username = input("Enter username :- ")

if username in dic :
    password = input("Enter password :- ")
    if dic[username] == password :
        print ("You are now logged into the system.")
    else :
        print ("Invalid password.")
else :
    print ("You are not valid user.")

Enter username :- Portal
Enter password :- 1912
You are now logged into the system.


# Program : 10

**Write a Python program to convert a given tuple of positive integers into an integer. Original tuple: (1, 2, 3) Convert the said tuple of positive integers into an integer: 123**


In [ ]:
def tuple_to_int(nums):
    result = int(''.join(map(str,nums)))
    return result

nums = (1,2,3)
print("Original tuple: ") 
print(nums)
print("Convert the said tuple of positive integers into an integer:")
print(tuple_to_int(nums))

Original tuple: 
(1, 2, 3)
Convert the said tuple of positive integers into an integer:
123


#Program : 11

**Write a Python program to sort a list of tuples using Lambda. Original list of tuples: – Original list of tuples: – [('Discrete Mathematics', 93), ('Python Programming', 95), ('Environmental Science', 88), ('Engineering Physics', 83)] – Sorting the List of Tuples: – [('Engineering Physics', 83), ('Environmental Science', 88), ('Discrete Mathematics', 93), ('Python Programming', 95)]**

In [ ]:
subject_marks = [('Discrete Mathematics', 93), ('Python Programming', 95), ('Environmental Science', 88), ('Engineering Physics', 83)] 
print("Original list of tuples:")
print(subject_marks)
subject_marks.sort(key = lambda x: x[1])
print("\nSorting the List of Tuples:")
print(subject_marks)

Original list of tuples:
[('Discrete Mathematics', 93), ('Python Programming', 95), ('Environmental Science', 88), ('Engineering Physics', 83)]

Sorting the List of Tuples:
[('Engineering Physics', 83), ('Environmental Science', 88), ('Discrete Mathematics', 93), ('Python Programming', 95)]


#Program : 12

**1.Write a program that inputs a tuple having words of a string as its element e.g., (“The”,” quick”,” brown”, “fox”) and then translate each text element to pig Latin.**

In [ ]:
tup = ("The", "quick", "brown", "fox")
index = 0
while index < len(tup):
    new = tup[ index ][1 : ] + tup[ index ][ 0 ] + "ay"
    print (new, end = " ")
    index += 1

heTay uickqay rownbay oxfay 

**2.Write a program that takes a tuple with words in pig latin (as created by previous question) and translate it back to original text.**

In [ ]:
tup =("heTay","uickqay","rownbay", "oxfay") 
index = 0
while index < len(tup):
    new = tup[ index ][ -3 ] + tup[ index ][0 : -3 ] 
    print (new, end = " ")
    index += 1

The quick brown fox 

#Program : 13

**Write a program to print Fibonacci Series using recursion.**

In [ ]:
def recur_fibo(n):  
   if n <= 1:  
       return n  
   else:  
       return(recur_fibo(n-1) + recur_fibo(n-2))  

nterms = int(input("How many terms? "))  

if nterms <= 0:  
   print("Plese enter a positive integer")  
else:  
   print("Fibonacci sequence:")  
   for i in range(nterms):  
       print(recur_fibo(i))  

How many terms? 5
Fibonacci sequence:
0
1
1
2
3


# Program : 14

**Write a menu-driven program to implement linear search and binary search.**

**Linear search**

In [ ]:
def LinearSearch(array, n, k):

    for j in range(0, n):

        if (array[j] == k):

            return j

    return -1

 
array = [1, 3, 5, 7, 9]

k = 7
n = len(array)

result = LinearSearch(array, n, k)

if(result == -1):

    print("Element not found")

else:

    print("Element found at index: ", result)

Element found at index:  3


**Binary Search**

In [ ]:
def binarySearch(arr, k, low, high):
   
    while low <= high:
        mid = low + (high - low)//2
        if arr[mid] == k:
            return mid
        elif arr[mid] < k:
            low = mid + 1
        else:
            high = mid - 1
    return -1

arr = [1, 3, 5, 7, 9]

k = 5
result = binarySearch(arr, k, 0, len(arr)-1)

if result != -1:
    print("Element is present at index " + str(result))
else:
    print("Not found")

Element is present at index 2


#Program : 15

**Write a program to implement merge sort, selection sort and insertion sort.**

**merge sort**

In [ ]:
def mergeSort(arr):
	if len(arr) > 1:

		mid = len(arr)//2

		L = arr[:mid]
		R = arr[mid:]

		mergeSort(L)
		mergeSort(R)

		i = j = k = 0

		while i < len(L) and j < len(R):
			if L[i] < R[j]:
				arr[k] = L[i]
				i += 1
			else:
				arr[k] = R[j]
				j += 1
			k += 1

		while i < len(L):
			arr[k] = L[i]
			i += 1
			k += 1

		while j < len(R):
			arr[k] = R[j]
			j += 1
			k += 1

def printList(arr):
	for i in range(len(arr)):
		print(arr[i], end=" ")
	print()

if __name__ == '__main__':
	arr = [12, 11, 13, 5, 6, 7]
	print("Given array is", end="\n")
	printList(arr)
	mergeSort(arr)
	print("Sorted array is: ", end="\n")
	printList(arr)

Given array is
12 11 13 5 6 7 
Sorted array is: 
5 6 7 11 12 13 


**selection sort**

In [ ]:
def selection_sort(nums):
  
    for i in range(len(nums)):
        lowestvalue = i

        for j in range(i + 1, len(nums)):
            if nums[j] < nums[lowestvalue]:
                lowestvalue = j
        
        nums[i], nums[lowestvalue] = nums[lowestvalue], nums[i]

random_list_of_nums = [12, 8, 3, 20, 11]
selection_sort(random_list_of_nums)
print(random_list_of_nums)

[3, 8, 11, 12, 20]


**insertion sort.**

In [ ]:
def insertion_sort(nums):
    
    for i in range(1, len(nums)):
        item_to_insert = nums[i]
        
        j = i - 1
        
        while j >= 0 and nums[j] > item_to_insert:
            nums[j + 1] = nums[j]
            j -= 1
        
        nums[j + 1] = item_to_insert

random_list_of_nums = [9, 1, 28, 6]
insertion_sort(random_list_of_nums)
print(random_list_of_nums)

[1, 6, 9, 28]
